sdc11073/sdcdevice/sdc_handlers.py

```diff
@@ -442,53 +471,39 @@ def sendDescriptorUpdates(self, mdibVersion, updated, created, deleted, updated_
-            for descriptorHandle, changedSample in changedSamples.items():
-                determinationTime = changedSample.determinationTime
-                samples = [s[0] for s in changedSample.samples]  # only the values without the 'start of cycle' flags
-                activationState = changedSample.activationState
-                st = tr.getRealTimeSampleArrayMetricState(descriptorHandle)
-                if st.metricValue is None:
-                    st.mkMetricValue()
-                st.metricValue.Samples = samples
-                st.metricValue.DeterminationTime = determinationTime  # set Attribute
-                st.metricValue.Annotations = changedSample.annotations
-                st.metricValue.ApplyAnnotations = changedSample.applyAnnotations
-                st.ActivationState = activationState
-
-    def _rtSampleSendLoop(self):
-        if PROFILING:
-            pr = cProfile.Profile()
-        time.sleep(
-            0.1)  # start delayed in order to have a fully initialized device when waveforms start (otherwise timing issues might happen)
+    def _rt_sample_sendloop(self):
+        """Periodically send waveform samples."""
+        # start delayed in order to have a fully initialized device when waveforms start
+        # (otherwise timing issues might happen)
+        time.sleep(0.1)
         timer = intervaltimer.IntervalTimer(periodInSeconds=self.collectRtSamplesPeriod)
-        if PROFILING:
-            pr_time = time.monotonic()
-            initial_time = pr_time  # delayed start of profiler, ignore init calls
         while self._runRtSampleThread:
-            if PROFILING:
-                if initial_time is not None and time.monotonic() - initial_time > 2:
-                    pr.enable()
-                    initial_time = None
             behindScheduleSeconds = timer.waitForNextIntervalBegin()
-            changedSamples = self._mdib.getUpdatedDeviceRtSamples()
-            if len(changedSamples) > 0:
-                self._logWaveformTiming(behindScheduleSeconds)  #
-                self.sendWaveformUpdates(changedSamples)
-            if PROFILING and initial_time is None:
-                if time.monotonic() - pr_time > 5:
-                    print('profile')
-                    pr.disable()
-                    s = StringIO()
-                    ps = pstats.Stats(pr, stream=s).sort_stats('time')
-                    ps.print_stats(30)
-                    print(s.getvalue())
-                    pr.enable()
-                    pr_time = time.monotonic()
+            self._mdib.update_all_rt_samples() # update from waveform generators
+            self._logWaveformTiming(behindScheduleSeconds)
+
+    def _periodic_reports_send_loop(self):
+        self._logger.debug('_periodic_reports_send_loop start')
+        time.sleep(0.1)  # start delayed
+        timer = intervaltimer.IntervalTimer(periodInSeconds=self._periodic_reports_interval)
+        while self._run_periodic_reports_thread:

```

Comment:
You really need to document somewhere that you're sending periodic reports with a constant periodicity, and do not care about what the mdib tells consumers about periodicity. Remember, every single metric can specify the periodicity with which it is send via periodic reports with the Retrievability element.